tools/btmidi-server.c

```diff
@@ -0,0 +1,887 @@
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_PGMCHANGE);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CHANPRESS);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_PITCHBEND);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SYSEX);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_QFRAME);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SONGPOS);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SONGSEL);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_TUNE_REQUEST);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CLOCK);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_START);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CONTINUE);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_STOP);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SENSING);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_RESET);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CONTROL14);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_NONREGPARAM);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_REGPARAM);
+
+	err = snd_seq_set_client_info(server->seq_handle, info);
+	if (err < 0) {
+		fprintf(stderr, "Could not set client info: %s (%d)\n", snd_strerror(err), err);
+		goto _err_port;
+	}
+
+	/* Input file descriptors */
+	snd_seq_poll_descriptors(server->seq_handle, &pfd, 1, POLLIN);
+
+	server->io = io_new(pfd.fd);
+	if (!server->io) {
+		fprintf(stderr, "Could not allocate I/O eventloop\n");
+		goto _err_port;
+	}
+
+	io_set_read_handler(server->io, midi_notify_cb, server, NULL);
+
+	//Init the MIDI parser
+	err = midi_read_init(&server->midi_in);
+	if (err < 0) {
+		fprintf(stderr, "Could not initialise MIDI input parser\n");
+		goto _err_port;
+	}
+
+	err = midi_write_init(&server->midi_out, bt_att_get_mtu(server->att) - 3);
+	if (err < 0) {
+		fprintf(stderr, "Could not initialise MIDI output parser\n");
+		goto _err_midi;
+	}
+
+	/* Add MIDI Service */
+	bt_string_to_uuid(&uuid, MIDI_UUID);
+	service = gatt_db_add_service(server->db, &uuid, true, 6);
+
+	/* Add MIDI IO Characteristics */
+	bt_string_to_uuid(&uuid, MIDI_IO_UUID);
+	server->midi_io = gatt_db_service_add_characteristic(service, &uuid,
+						BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+						BT_GATT_CHRC_PROP_WRITE_WITHOUT_RESP | BT_GATT_CHRC_PROP_READ | BT_GATT_CHRC_PROP_NOTIFY,
+						NULL, midi_write_cb, server);
+	server->midi_io_handle = gatt_db_attribute_get_handle(server->midi_io);
+
+	/* Add MIDI CCC */
+	bt_uuid16_create(&uuid, GATT_CLIENT_CHARAC_CFG_UUID);
+	gatt_db_service_add_descriptor(service, &uuid,
+					BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+					midi_ccc_read_cb, midi_ccc_write_cb, server);
+
+	/* Add MIDI Characteristic at 0x000b */
+	bt_uuid16_create(&uuid, GATT_CHARAC_UUID);
+	gatt_db_service_insert_characteristic(service, 0x000b, &uuid,
```

Comment:
I forgot about this.  This was something really weird i could not seem to figure out.  And I think this just stems from my insufficient GATT knowledge.  My testing was done with an IPAD and I was initially unable to connect to my server.  I used hcidump to see what was going on and it was looking for the Main MIDI Characteristic 0x2803.  From reading it appears that this should be defined for easy discovering and then a second characteristic is created with the actual value.

So what i did was the normal add characteristic for 0x2803, but it did not place it at handle 0x000b and based of hcidump the ipad insisted on it being at 0x000b.  So i used insert instead to put it at that handle.  This is where I could use some advice on exactly how to get these characteristics inserted.  There must be some link between them or something I am missing that would allow me to avoid manually forcing this characteristic to be at 0x000b